pkg/native/serializer.go

```diff
@@ -1,13 +1,21 @@
 package native
 
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 -generate
+
 import (
 	"io"
 
 	"github.com/bom-squad/protobom/pkg/sbom"
-	"github.com/bom-squad/protobom/pkg/writer/options"
 )
 
+//counterfeiter:generate . Serializer
 type Serializer interface {
-	Serialize(options.Options, *sbom.Document) (interface{}, error)
-	Render(options.Options, interface{}, io.Writer) error
+	Serialize(*sbom.Document, *SerializeOptions) (interface{}, error)
+	Render(interface{}, io.Writer, *RenderOptions) error
+}
+
+type SerializeOptions struct{}
+
+type RenderOptions struct {
+	Indent int
 }
```

Comment:
Those are redundant IMHO, and I think NewSerializer should include those options in instantiations (with or without options pattern), making Serializer simpler with only 

```go
type Serializer interface {
	Serialize(*sbom.Document) (interface{}, error)
	Render(interface{}, io.Writer) error
}
```

The only case against, is where you want to reconfigure a serializer forcing you to:
1) instantiate new serializer
2) register the new serializer with the writer

Still not sure the best route here as we don't "really" have options supported, planned.

Reply:
I disagree with this idea on two fundamental grounds.

First, the serializers and unserializers should behave on each call according to the specifications sent by the caller. This means taking the execution specification (the options) on every run, they should be as stateless as possible. For example, if I have an application running, say a server responding to serialization calls, I want it to behave according to the parameters sent by the client. If the client needs to render using a set of options on one call and a different set of options on the next one, the application needs to react without having to destroy an reallocate the (un)serializer. Creating and destroying these drivers as a response to a change of the execution parameteres seems like an anti pattern.

Second, the `WithXXX()`/`WithoutXXX()` pattern: I can sympathize with the idea of preconfiguring the object at creation time. Unfortunately, the `With..` pattern does not fit this architecture. Options received by the library can be potentially piped down to the serializers and this makes it impossible to apply it to this scenario. The main problem is that you need to write a method to enable each option They may not be generalized, or they may even be unknown. Let me expand this idea with an example in the writer/serializer combo:

General options that apply to the writer or to all serializers can be implemented using `WithX...` , no problem. For example `WithIndent(4)`: Indentation level is an option that can apply to all json encodings and it is mostly harmless if it gets piped to serializers whose encodings don't have indentation (say SPDX tag-value). The problem, however, arises when considering options that apply to just one serializer, or (worse) when thinking about third-party serializers that get added as plug-ins. 

Consider an option that only applies to cyclonedx 1.4. Say you want an option to die when (cdx 1.5) Lifecycles are defined. In this case, you'd need an option `WithDieOnLifcycle()`. But this applies only to cdx 1.4. The method needs to exist, but it is not general for all serializers. Where do you define it? On the writer? On the general options? On the serializer?

And even worse, this option is probably something we will never have on the "stock" serializers. But suppose my employer wants to write a custom CDX serializer that has this option and use it as a plugin. Then the method needs to exist but there is no easy way for third party serializers to define these methods to make them available when you create the writer.

This is where empty interfaces (or generics, but let us not go into that just yet) come in handy. When passing an options struct on each run, the drivers can check an options set and check if a field is a known type and configure the run.

Reply:
I agree some options can be given on the instance creation,
I don't have a strong opnion about this but let's consider If we expect the library user to use a single seralizer instance multiple times with different options.

Reply:
Right, this is the same take I have on this (see the comment on my review).

Reply:
It would be more logical to relocate the Serializer interface to the writer package. 
I just don't want to make the writer/native/serializer tightly coupled

Reply:
Also, I want to use a DTO if we pass "config" to the writer. I think serializer should not accept the writer config (cyclic dependency) but the writer should create the Render/SerializeOptions from the config.

Reply:
https://github.com/bom-squad/protobom/compare/refactor-core-serializers...demo-writer-with-config?expand=1

I pushed this new branch to showcase the general idea. `demo-writer-with-config`

what changed:
Writer now has default config (on instantiation)
Writer now have new functions `WriteStreamWithConfig` and `WriteFileWithConfig` to allow setting those configs on the fly.
Writer existing function uses default config
This config is converted to Render/SerializeOptions using DTOs (writer.Config -> native.RenderOption)

Reply:
There is still an issue with passing arbitrary options to serializers, but this is OOS in this PR.

Reply:
> if want us to pass arbitrary options to known/unknown parsers, option like `DieOnLifcycle` (CDX parser option) We need to either add "DieOnLifecycle" to "SerializeOptions" struct, make it an interface{} or god forbid use generics 😂
> 

Regarding this one, so if you think the best way to support an unknown prop bag for a serializer is to pass as an interface{} , that's fine by me. but the writer would have to accept a Config with 

```go
SerializeOptions interface{}
RenderOptions interface{}
```
and the writer will delegate those to their respective function calls.